src/scope/future/mod.rs

```diff
@@ -0,0 +1,464 @@
+#[allow(warnings)]
+use log::Event::*;
+use futures::{Async, Future, Poll};
+use futures::executor;
+use futures::future::CatchUnwind;
+use futures::task::{self, Spawn, Task, Unpark};
+use job::{Job, JobRef};
+use registry::{Registry, WorkerThread};
+use std::any::Any;
+use std::panic::AssertUnwindSafe;
+use std::mem;
+use std::sync::Arc;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::*;
+use std::sync::Mutex;
+use unwind;
+
+use super::Scope;
+
+const STATE_PARKED: usize = 0;
+const STATE_UNPARKED: usize = 1;
+const STATE_EXECUTING: usize = 2;
+const STATE_EXECUTING_UNPARKED: usize = 3;
+const STATE_COMPLETE: usize = 4;
+
+// Warning: Public end-user API.
+pub struct RayonFuture<T, E> {
+    inner: Arc<ScopeFutureTrait<Result<T, E>, Box<Any + Send + 'static>>>,
+}
+
+/// Create a `RayonFuture` that will execute `F` and yield its result,
+/// propagating any panics.
+///
+/// Unsafe because caller asserts that all references in `F` will
+/// remain valid at least until `counter` is decremented via `set()`.
+/// In practice, this is ensured by the `scope()` API, which ensures
+/// that `F: 'scope` and that `'scope` does not end until `counter`
+/// reaches 0.
+///
+/// NB. This is a free fn so that we can expose `RayonFuture` as public API.
+pub unsafe fn new_rayon_future<'scope, F>(future: F,
+                                          scope: *const Scope<'scope>)
+                                          -> RayonFuture<F::Item, F::Error>
+    where F: Future + Send + 'scope,
+{
+    // We always have to have a non-null scope (for now, anyway):
+    debug_assert!(!scope.is_null());
+
+    let inner = ScopeFuture::spawn(future, scope);
+    return RayonFuture { inner: hide_lifetime(inner) };
+
+    unsafe fn hide_lifetime<'l, T, E>(x: Arc<ScopeFutureTrait<T, E> + 'l>)
+                                      -> Arc<ScopeFutureTrait<T, E>> {
+        mem::transmute(x)
+    }
+}
+
+impl<T, E> RayonFuture<T, E> {
+    pub fn rayon_wait(mut self) -> Result<T, E> {
+        let worker_thread = WorkerThread::current();
+        if worker_thread.is_null() {
+            self.wait()
+        } else {
+            // Assert that uses of `worker_thread` pointer below are
+            // valid (because we are on the worker-thread).
+            unsafe {
+                (*worker_thread).wait_until(&*self.inner);
+                debug_assert!(self.inner.probe());
+                self.poll().map(|a_v| match a_v {
+                    Async::Ready(v) => v,
+                    Async::NotReady => panic!("probe() returned true but poll not ready")
+                })
+            }
+        }
+    }
+}
+
+impl<T, E> Future for RayonFuture<T, E> {
+    type Item = T;
+    type Error = E;
+
+    fn wait(self) -> Result<T, E> {
+        if WorkerThread::current().is_null() {
+            executor::spawn(self).wait_future()
+        } else {
+            panic!("using  `wait()` in a Rayon thread is unwise; try `rayon_wait()`")
```

Comment:
Is there a way to make `RayonFuture::wait` a compile-time error, rather than a panic?  Probably not, since we don't control the trait, but it's ugly that this will only show up at runtime.